posts/2022-10-28-gats-stabilization.md

```diff
@@ -0,0 +1,275 @@
+
+### Non-local requirements for where clauses on GATs
+
+The last limitation we'll talk about today is a bit different than the others; it's not a bug and it shouldn't prevent any programs from compiling. But it all comes back to that `where Self: 'a` clause you've seen in several parts of this post. As mentioned before, if you're interested in digging a bit into why that clause is required, see the [push for stabilization post][stabilization_post].
+
+There is one not-so-ideal about this clause: you must write it on the trait. Like with where clauses on functions, you cannot add clauses to associated types in impls that aren't there in the trait. However, if you *didn't* add this clause, a large set of potential impls of the trait would be disallowed.
+
+To help users not fall into the pitfall of accidentally forgetting to add this (or similar clauses that end up with the same effect for a different set of generics), we've implemented a set of rules that must be followed for a trait with GATs to compile. Let's first look at the error without writing the clause:
+
+```rust
+trait LendingIterator {
+    type Item<'a>;
+
+    fn next<'a>(&'a mut self) -> Self::Item<'a>;
+}
+```
+
+```rust
+error: missing required bound on `Item`
+ --> src/lib.rs:2:5
+  |
+2 |     type Item<'a>;
+  |     ^^^^^^^^^^^^^-
+  |                  |
+  |                  help: add the required where clause: `where Self: 'a`
+  |
+  = note: this bound is currently required to ensure that impls have maximum flexibility
+  = note: we are soliciting feedback, see issue #87479 <https://github.com/rust-lang/rust/issues/87479> for more information
+```
+
+This error should hopefully be helpful (you can even `cargo fix` it!). But, what exactly are these rules? Well, ultimately, they end up being somewhat simple: *for methods that use the GAT, any bounds that can be proven must also be present on the GAT itself*.
+
+Okay, so how did we end up with the required `Self: 'a` bound. Well, let's take a look at the `next` method. It returns `Self::Item<'a>`, and we have an argument `&'a mut self`. We're getting a bit into the details of the Rust language, but because of that argument, we know that `Self: 'a` must hold. So, we require that bound.
+
+We're requiring these bounds now to leave room in the future to potentially imply these automatically (and of course because it should help users writing traits with GATs). They shouldn't interfere with any real use-cases, but if you do encounter a problem, check out the issue mentioned in the error abovee. And if you want to see a fairly comprehensive testing of different scenarios on what bounds are required and when, check out the relevant [test file](https://github.com/rust-lang/rust/blob/f2702e922ba31e49d6167f5651d4545646dcf22d/src/test/ui/generic-associated-types/self-outlives-lint.rs).
```

Comment:
```suggestion
We're requiring these bounds now to leave room in the future to potentially imply these automatically (and of course because it should help users write traits with GATs). They shouldn't interfere with any real use-cases, but if you do encounter a problem, check out the issue mentioned in the error above. And if you want to see a fairly comprehensive testing of different scenarios on what bounds are required and when, check out the relevant [test file](https://github.com/rust-lang/rust/blob/f2702e922ba31e49d6167f5651d4545646dcf22d/src/test/ui/generic-associated-types/self-outlives-lint.rs).
```